src/box/field_default_func.c

```diff
@@ -0,0 +1,113 @@
+/*
+ * SPDX-License-Identifier: BSD-2-Clause
+ *
+ * Copyright 2010-2023, Tarantool AUTHORS, please see AUTHORS file.
+ */
+#include "port.h"
+#include "engine.h"
+#include "tt_static.h"
+#include "tuple_format.h"
+#include "field_default_func.h"
+
+/**
+ * Check `func` for compliance with the field default function rules.
+ *
+ * Returns 0 on success. On error, sets diag and returns -1.
+ */
+static int
+field_default_func_verify(struct func *func)
+{
+	const char *func_name = func->def->name;
+
+	if (func->def->language == FUNC_LANGUAGE_LUA) {
+		if (func->def->body != NULL)
+			return 0;
+		diag_set(ClientError, ER_CREATE_DEFAULT_FUNC, func_name,
+			 "Lua function must have persistent body");
+		return -1;
+	}
+
+	/* TODO(gh-8793): Support SQL */
+	diag_set(ClientError, ER_CREATE_DEFAULT_FUNC, func_name,
+		 "unsupported language");
+	return -1;
+}
+
+/**
+ * Implementation of field_default_func_call().
+ * Called by pointer to avoid linking dependencies.
+ */
+static int
+field_default_func_call_impl(struct field_default_func *default_func,
+			     const char *arg_data, uint32_t arg_size,
+			     const char **ret_data, uint32_t *ret_size)
+{
+	struct port out_port, in_port;
+	port_c_create(&in_port);
+	if (arg_data != NULL)
+		port_c_add_mp(&in_port, arg_data, arg_data + arg_size);
+
+	int rc = func_call_no_access_check(default_func->holder.func,
+					   &in_port, &out_port);
+	port_destroy(&in_port);
+	if (rc != 0) {
+		diag_set(ClientError, ER_DEFAULT_FUNC_FAILED,
+			 default_func->holder.func->def->name,
+			 diag_last_error(diag_get())->errmsg);
+		return -1;
+	}
+
+	*ret_data = port_get_msgpack(&out_port, ret_size);
+	assert(mp_typeof(**ret_data) == MP_ARRAY);
+	unsigned ret_count = mp_decode_array(ret_data);
+	if (ret_count != 1) {
+		diag_set(ClientError, ER_DEFAULT_FUNC_FAILED,
+			 default_func->holder.func->def->name,
+			 tt_sprintf("expected 1 return value, got %u",
+				    ret_count));
+		return -1;
```

Comment:
It seem that you must `port_destroy(&out_port);` in case of this error. I think that you should call it right after `port_get_msgpack`

Reply:
Fixed.

Also rebased on top of #8930 and renamed `space_reattach_defaults`/`space_detach_defaults` to `space_pin_defaults`/`space_unpin_defaults` to match `space_pin_collations`/`space_unpin_collations`.